File: monix-execution/jvm/src/main/java/monix/execution/atomic/boxes/left64Java7/BoxedLong.java
/*
 * Copyright (c) 2016 by its authors. Some rights reserved.
 * See the project homepage at: https://sincron.org
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package monix.execution.atomic.boxes.left64Java7;

import monix.execution.atomic.boxes.common.LeftPadding48;
import monix.execution.misc.UnsafeAccess;
import java.lang.reflect.Field;

public final class BoxedLong extends LeftPadding48
        implements monix.execution.atomic.boxes.BoxedLong {

    public volatile long value;
    public static final long OFFSET;
    static {
        try {
            Field field = BoxedLong.class.getDeclaredField("value");
            OFFSET = UnsafeAccess.UNSAFE.objectFieldOffset(field);
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    public BoxedLong(long initialValue) {
        this.value = initialValue;
    }

    public long volatileGet() {
        return value;
    }

    public void volatileSet(long update) {
        value = update;
    }

    public void lazySet(long update) {
        UnsafeAccess.UNSAFE.putOrderedLong(this, OFFSET, update);
    }

    public boolean compareAndSet(long current, long update) {
        return UnsafeAccess.UNSAFE.compareAndSwapLong(this, OFFSET, current, update);
    }

    public long getAndSet(long update) {
        long current = value;
        while (!UnsafeAccess.UNSAFE.compareAndSwapLong(this, OFFSET, current, update))
            current = value;
        return current;
    }
}

File: monix-execution/shared/src/test/scala/monix/execution/atomic/BoxedLong.scala
/*
 * Copyright (c) 2014-2018 by The Monix Project Developers.
 * See the project homepage at: https://monix.io
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package monix.execution.atomic

case class BoxedLong(value: Long)

object BoxedLong {
  final val MinValue = BoxedLong(Long.MinValue)
  final val MaxValue = BoxedLong(Long.MaxValue)

  implicit val numeric = new Numeric[BoxedLong] {
    def plus(x: BoxedLong, y: BoxedLong): BoxedLong =
      BoxedLong(x.value + y.value)
    def toDouble(x: BoxedLong): Double =
      x.value.toDouble
    def toFloat(x: BoxedLong): Float =
      x.value.toFloat
    def toInt(x: BoxedLong): Int =
      x.value.toInt
    def negate(x: BoxedLong): BoxedLong =
      BoxedLong(-1 * x.value)
    def fromInt(x: Int): BoxedLong =
      BoxedLong(x)
    def toLong(x: BoxedLong): Long =
      x.value
    def times(x: BoxedLong, y: BoxedLong): BoxedLong =
      BoxedLong(x.value * y.value)
    def minus(x: BoxedLong, y: BoxedLong): BoxedLong =
      BoxedLong(x.value - y.value)
    def compare(x: BoxedLong, y: BoxedLong): Int =
      x.value.compareTo(y.value)
    def parseString(str: String): Option[BoxedLong] =
      try Some(BoxedLong(str.toLong))
      catch { case _: NumberFormatException => None }
  }
}
